pallets/liquidity-provider/src/transfer_handlers.rs

```diff
@@ -0,0 +1,164 @@
+    to: <T as frame_system::Trait>::AccountId,
+    to_amount: Balance<T>,
+) -> dispatch::DispatchResult
+where
+    T: Trait,
+{
+    let orig = collateral_balances::<T>(&from);
+    let mut modified = collateral_balances::<T>(&from);
+    let actually_transferred = if let Some(rslt) = test_transfer_evenly(&mut modified, to_amount) {
+        rslt
+    } else {
+        return Err(crate::Error::<T>::NoFunds.into());
+    };
+    do_transfer_evenly(actually_transferred, from, modified, orig, to)?;
+    Ok(())
+}
+
+struct Account<T>
+where
+    T: Trait
+{
+    asset: <<T as Trait>::Collateral as MultiCurrency<<T as frame_system::Trait>::AccountId>>::CurrencyId,
+    reserved: Balance<T>
+}
+
+// Checks `wants_to_transfer` and returns the sum of all collaterals
+#[inline]
+fn check_if_usdv_to_transfer_is_greater_than_total_collaterals<T>(
+    collateral_balances: &Balances<T>,
+    wants_to_transfer: Balance<T>,
+) -> Option<Balance<T>>
+where
+    T: Trait,
+{
+    let mut total_collaterals: Balance<T> = Balance::<T>::zero();
+
+    for Account { reserved, .. } in collateral_balances.iter() {
+        total_collaterals = total_collaterals.checked_add(reserved)?;
+    }
+
+    if wants_to_transfer > total_collaterals {
+        return None;
+    }
+
+    Some(total_collaterals)
+}
+
+// Creates a snapshot with all non-empty collateral balances of a given user
+#[inline]
+fn collateral_balances<T>(who: &<T as frame_system::Trait>::AccountId) -> Balances<T>
+where
+    T: Trait,
+{
+    Collateral::variants()
+        .iter()
+        .filter_map(|&collateral| {
+            let collateral_asset = Asset::Collateral(collateral);
+            let reserved = T::Collateral::reserved_balance(collateral_asset, who);
+
+            if reserved == Balance::<T>::zero() {
+                None
+            } else {
+                Some(Account {
+                    asset: collateral_asset,
+                    reserved,
+                })
+            }
+        })
+        .collect()
+}
+
+// Takes the original and subtracted balances and issue transfer operations
+#[inline]
+fn do_transfer_evenly<T>(
+    amount: Balance<T>,
+    from: <T as frame_system::Trait>::AccountId,
+    modified: Balances<T>,
+    original: Balances<T>,
+    to: <T as frame_system::Trait>::AccountId,
+) -> dispatch::DispatchResult
+where
+    T: Trait,
+{
+    for (a, b) in modified.iter().zip(original.iter()) {
+        let diff = b
+            .reserved
+            .checked_sub(&a.reserved)
+            .ok_or(crate::Error::<T>::NoFunds)?;
+        T::Collateral::repatriate_reserved(a.asset, &from, &to, diff, BalanceStatus::Reserved)?;
+    }
+    T::Asset::transfer(Asset::Usdv, &from, &to, amount)?;
+    Module::<T>::deposit_event(RawEvent::Transfer(from, to, amount));
+    Ok(())
+}
+
+// Just check if a given set of collateral balances can fulfil the desired amount of USDv by
+// subtracting each asset
+#[inline]
+fn test_transfer_evenly<T>(
```

Comment:
Since we mint USDv 1:1 with the collateral, would it be safe to assume USDv balance is in sync with its collateral, so it would be enough to check at the beginning if I can transfer the USDv then I can transfer the collateral?

Reply:
The additional checking is a plus since this function is still necessary to correctly calculate how much each collateral will be subtracted.